game/gamestate.go

```diff
@@ -147,3 +539,263 @@ func (gs *GameState) AreAdjacent(cantonID1, cantonID2 int) bool {
 	}
 	return false
 }
+
+func (gs *GameState) Hash() StateHash {
+	hasher := fnv.New64a()
+
+	// Hash current player
+	binary.Write(hasher, binary.LittleEndian, int64(gs.CurrentPlayer))
+
+	// Hash phase
+	binary.Write(hasher, binary.LittleEndian, int64(gs.Phase))
+
+	// Hash troop counts
+	for _, count := range gs.TroopCounts {
+		binary.Write(hasher, binary.LittleEndian, int64(count))
+	}
+
+	// Hash ownerships
+	for _, owner := range gs.Ownership {
+		binary.Write(hasher, binary.LittleEndian, int64(owner))
+	}
+
+	// Hash TroopsToPlace
+	binary.Write(hasher, binary.LittleEndian, int64(gs.TroopsToPlace))
+
+	// Hash PlayerTroops
+	for playerID, troops := range gs.PlayerTroops {
+		binary.Write(hasher, binary.LittleEndian, int64(playerID))
+		binary.Write(hasher, binary.LittleEndian, int64(troops))
+	}
+
+	return StateHash(hasher.Sum64())
+}
+
+// Just BFS
+func (gs *GameState) AreConnected(fromID, toID, playerID int) bool {
+	if fromID == toID {
+		return true
+	}
+	visited := make(map[int]bool)
+	queue := []int{fromID}
+
+	for len(queue) > 0 {
+		current := queue[0]
+		queue = queue[1:]
+
+		if visited[current] {
+			continue
+		}
+		visited[current] = true
+
+		for _, adjID := range gs.Map.Cantons[current].AdjacentIDs {
+			if gs.Ownership[adjID] != playerID {
+				continue
+			}
+			if adjID == toID {
+				return true
+			}
+			if !visited[adjID] {
+				queue = append(queue, adjID)
+			}
+		}
+	}
+	return false
+}
+
+func (gs *GameState) Play(move Move) State {

```

Comment:
- [ ] this method should have value receiver (not mutating the state object and returns a new state)
```suggestion
func (gs GameState) Play(move Move) State {
```